addons/mail/models/mail_thread.py

```diff
@@ -2629,37 +2630,51 @@ def _notify_classify_recipients(self, recipient_data, model_name, msg_vals=None)
 
-        if model_name:
-            view_title = _('View %s', model_name)
-        else:
-            view_title = _('View')
-
         # fill group_data with default_values if they are not complete
-        for group_name, group_func, group_data in groups:
-            group_data.setdefault('notification_group_name', group_name)
-            group_data.setdefault('notification_is_customer', False)
-            is_thread_notification = self._notify_get_recipients_thread_info(msg_vals=msg_vals)['is_thread_notification']
-            group_data.setdefault('has_button_access', is_thread_notification)
-            group_button_access = group_data.setdefault('button_access', {})
-            group_button_access.setdefault('url', access_link)
-            group_button_access.setdefault('title', view_title)
-            group_data.setdefault('actions', list())
-            group_data.setdefault('recipients', list())
+        for lang, lang_groups in groups_by_lang.items():
+            for group_name, group_func, group_data in lang_groups:
+                context = {'lang': lang}
+                model_name = (model_id.with_context(lang=lang).display_name or model_name) if model_id else model_name
+                if model_name:
+                    view_title = _('View %s', model_name)
+                else:
+                    view_title = _('View')
+                group_data.setdefault('notification_group_name', group_name)
+                group_data.setdefault('notification_is_customer', False)
+                is_thread_notification = self._notify_get_recipients_thread_info(msg_vals=msg_vals)['is_thread_notification']
+                group_data.setdefault('has_button_access', is_thread_notification)
+                group_button_access = group_data.setdefault('button_access', {})
+                group_button_access.setdefault('url', access_link)
+                group_button_access.setdefault('title', view_title)
+                group_data.setdefault('actions', list())
+                group_data.setdefault('recipients', list())
+                del context
 
         # classify recipients in each group
-        for recipient in recipient_data:
-            for group_name, group_func, group_data in groups:
-                if group_func(recipient):
-                    group_data['recipients'].append(recipient['id'])
-                    break
+        for lang, recipient_ids in [(group['lang'], group['ids']) for group in recipient_lang_groups]:
+            for recipient_id in recipient_ids:
+                for group_name, group_func, group_data in groups_by_lang[lang]:
+                    if group_func(partner_id_to_recipient_data[recipient_id]):
+                        group_data['recipients'].append(recipient_id)
+                        break
 
         result = []
-        for group_name, group_method, group_data in groups:
+        for group_name, dummy, group_data in [
```

Comment:
If I am not wrong, this changes the returned value of "_notify_classify_recipients", and groups are not unique anymore ... which is annoying in stable (and not aligned with master I think, where we first classify by lang then classify by group if I am not wrong)